### PR TITLE
add sec-ch-ua to chromium based user-agents, add count param to api

### PR DESCRIPTION
**Pull request by aabdulbasset**
## Description
- Added "?count=10" to api params.
- Add sec-ch-x to headers if a chromium user agent was selected.
- Copy headers_data instead of referencing it to allow per-request change.
- Add SEC_CH_UA to assets.
## Testing
- All tests passed locally.
## Misc
- Will need documentation changes.
- default api response is now an array instead of an object, can possibly be changed.
- Can possibly use a user-agent parser instead of manually identifying user agent.
- Need an automated way/api to populate sec_ch_ua array in assets

### CODE REVIEW

File: src/api/routes.py
Comment (essteer) [inline src/api/routes.py:1]: Good suggestion, thanks.

Please submit this as a separate PR after merging the latest changes from `main` &mdash; the `count` param is thematically different to the rest of the PR and will be a breaking change due to the return object change from `dict` to `list[dict]`.

File: src/masquer/utils/assets.py
Comment (essteer) [inline src/masquer/utils/assets.py:1]: This file is overwritten when `update.py` (in the project root) is executed.

A quick fix would be to hardcode these values into `update.py`, though as you mention in the PR description an automated method would be best.

File: src/masquer/utils/sec_ch.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+def get_browser(user_agent: str) -> str:
+    if('Edg' in user_agent):
+        return 'Microsoft Edge'
+    elif('OPR' in user_agent):
+        # TODO: Opera have a custom sec-ch-ua for mobile version...
+        return 'Opera'
+    else:
+        return 'Google Chrome'
+
+def get_platform(user_agent: str) -> str:
```
Comment (essteer) [inline src/masquer/utils/sec_ch.py:29]: We could refactor this to be a bit cleaner, e.g.:

```py
def get_platform(useragent: str) -> str:
    platforms = ["Android", "iPhone", "Linux", "Macintosh", "Windows"]
    for platform in platforms:
        if platform in useragent:
            return platform
    return "Windows"
```

NB: "useragent" is treated as one word elsewhere in the programme, so the same format is applied here to be consistent.

File: src/masquer/utils/sec_ch.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    if version in SEC_CH_UA:
+        return SEC_CH_UA[version]
+    else:
+        return "Not Found"
+    
+def get_browser(user_agent: str) -> str:
+    if('Edg' in user_agent):
+        return 'Microsoft Edge'
+    elif('OPR' in user_agent):
+        # TODO: Opera have a custom sec-ch-ua for mobile version...
```
Comment (essteer) [inline src/masquer/utils/sec_ch.py:24]: Noted

**Review by essteer — CHANGES_REQUESTED** [review]
Hi @aabdulbasset &mdash; thanks for these suggested changes, much appreciated and you're welcome to contribute.

I've made a few review comments: adding the `count` param and introducing `sec-ch-ua` are both good ideas but I suggest splitting them into separate PRs &mdash; the extra param can probably go straight into the next release.

On the `sec-ch-ua` feature: I'd like to check first whether there's a source for updated attributes for this as per the user-agent and referer data before integrating. That will involve a few minor changes to things like the update script as well.

Note that a separate PR was merged since you submitted this. It should have little impact on the files in this PR but should be merged beforehand.  The docs will be cleaned up / improved soon anyway, I can add content for these features when needed.

If you're happy / have time to continue on this then please do when convenient, otherwise I can take it from here.

Thanks again!

File: src/masquer/utils/assets.py
Comment (aabdulbasset) [reply src/masquer/utils/assets.py:1]: Hmm, what do you think of using `https://github.com/fa0311/latest-user-agent/blob/main/header.json` ?
Seems to be conveniently updated every major browser release.

Comment (essteer) [reply src/masquer/utils/assets.py:1]: That seems like a decent source. I'm thinking about how to maintain the same approach of weighted selections based on the most-used values. One way could be to look up the `sec-ch-ua` value based on the selected user-agent if a Chromium user-agent is chosen.

For example if this user-agent is selected:

```py
"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0.0.0 Safari/537.3"
```

Parse the `Chrome/128` and map this to the first entry from your dict:

```py
SEC_CH_UA = {
    'chrome/128': '"Chromium";v="128", "Not;A=Brand";v="24", "%c";v="128"',
    'chrome/127': '"Not)A;Brand";v="99", "%c";v="127", "Chromium";v="127"',
}
```

If the value of `Not A Brand` was irrelevant then a new `sec-ch-ua` could be generated for each Chrome user-agent. The versions seem to be fixed though, so presumably browsers won't accept any random number there.

That might mean going back through previous versions of that repo or another source for an initial list &mdash; the current group of user-agents includes a Chromium release as far back as `Chrome/109` &mdash; then updating over time with new releases.

There doesn't seem to be information on the most used `sec-ch-ua` values in the same way there is for other header data, so there's also the question of whether it's significant or just using a handful of recent versions would be fine.

Comment (aabdulbasset) [reply src/masquer/utils/assets.py:1]: First approach is the current implemented one I believe. Did I misunderstand something ? 

Indeed `Not A Brand` is important, it changes every major browser version, with the main changes being the location and the special symbols. 
A good reference: https://wicg.github.io/ua-client-hints/#grease

Some websites depend on those browser hints way more than they should in identifying the user, since these are not random and cannot be statistically generated. But I think since most users are on the latest version or two, a handful or versions would be enough.

File: src/masquer/utils/sec_ch.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
+from .assets import (SEC_CH_UA, HEADER_DATA)
+import re
+def append_sec_ch(user_agent: str, headers: dict) -> str:
+    if(user_agent.find('Chrome') == -1):
+        return headers
+    platform = get_platform(user_agent)
+    headers['sec-ch-ua-platform'] = platform
+    headers['sec-ch-ua-mobile'] = "?1" if is_mobile(user_agent) else "?0"
+    headers['sec-ch-ua'] = find_sec_ch(get_chrome_version(user_agent)).replace('%c', get_browser(user_agent))
```
Comment (essteer) [inline src/masquer/utils/sec_ch.py:9]: I may have read this incorrectly, but it seems that if "Not Found" is returned by `find_sec_ch` because the version doesn't exist in the `SEC_CH_UA` dict, that will be the final value for `sec-ch-ua` in the header?

If that's correct, is it valid to send this? 

Whether it is or isn't valid, perhaps the approach should be to first check whether the Chromium version is in the dict before adding any of these attributes &mdash; if not, then revert to the default headers as per lines 4 & 5?

It takes me a bit of time to keep the values in my head with method chaining, please consider unpacking that last line into separate variables:

```py
browser = get_browser(useragent)
version = get_chrome_version(useragent)
sec_ch_ua = find_sec_ch(version)
sec_ch_ua = sec_ch_ua.replace('%c', browser)
headers['sec-ch-ua'] = sec_ch_ua
```

File: src/masquer/utils/assets.py
Comment (essteer) [reply src/masquer/utils/assets.py:1]: Thanks for that link - and yes you're right about the implementation, excuse my oversight.

The versions in the repo you shared go back as far as Chrome/124 &mdash; 121 and 109 both seem to be in common use (109 due to use by Amazon Cloudfront) so I've added those in below also.

```py
SEC_CH_UA = {
    'chrome/129': '"%c";v="129", "Not=A?Brand";v="8", "Chromium";v="129"',
    'chrome/128': '"Chromium";v="128", "Not;A=Brand";v="24", "%c";v="128"',
    'chrome/127': '"Not)A;Brand";v="99", "%c";v="127", "Chromium";v="127"',
    'chrome/126': '"Not/A)Brand";v="8", "Chromium";v="126", "%c";v="126"',
    'chrome/125': '"%c";v="125", "Chromium";v="125", "Not.A/Brand";v="24"',
    'chrome/124': '"Chromium";v="124", "%c";v="124", "Not-A.Brand";v="99"',
    'chrome/121': '"Not A(Brand";v="99", "%c";v="121", "Chromium";v="121"',
    'chrome/109': '"Not_A Brand";v="99", "%c";v="109", "Chromium";v="109"'
}
```

We could go with this for now and just load it in the same way the default header data is loaded.

Comment (essteer) [reply src/masquer/utils/assets.py:1]: A couple of things:

1) `sec-ch-ua` is still regarded as an experimental feature ([MDN docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Sec-CH-UA)), so the safe approach might be to stick with `/masq` for the current behaviour, and add another endpoint that mixes in the `sec-ch-ua` feature for Chromium browsers. 

A benefit of this is that if for whatever reason there is a use-case to not send  `sec-ch-ua`, then that option remains open.

2) Currently the data in use here is all for desktop user-agents, so the `is_mobile()` function will always return `False` and no iPhone or Android user-agents will be featured.

Another endpoint could be added (later) that provides the option to use either mobile or desktop user-agents, or a mix of both.

File: src/masquer/utils/sec_ch.py (hunk 3 above)
Comment (aabdulbasset) [reply src/masquer/utils/sec_ch.py:9]: It is correct, but neither this nor completely removing it will be valid, so I thought this might be a viable solution if the anti-bot solution is just checking for the header existence instead of checking if it has the correct value.

as far as I know, Amazon Cloudfront(as an example) has some kind of self-training, where they start by allowing the request if the header just exists, and starts getting more strict as the chrome version gets older/user IP makes multiple requests using same set of headers.

Sure thing, Sorry for that, I'm not good at python.

Comment (essteer) [reply src/masquer/utils/sec_ch.py:9]: Thanks - the code is fine, it's just my limited reading ability. 😄

OK noted on the header checks, that's really interesting. I'll read more into this. If the main `sec-ch-ua` are captured in the options then hopefully it should rarely pose a problem anway.

**Comment by essteer** [comment]
@aabdulbasset let me know if you're interested in making the changes discussed? There's no rush but if you're keen then I'll wait for your update, otherwise I'll pull your branch and continue from there sometime soon. Thanks!

**Comment by aabdulbasset** [comment]
> @aabdulbasset let me know if you're interested in making the changes discussed? There's no rush but if you're keen then I'll wait for your update, otherwise I'll pull your branch and continue from there sometime soon. Thanks!

sorry for being late ! had too much work to finish
will probably push the 2 PRs today/tomorrow

**Comment by essteer** [comment]
Ah, great stuff! I wasn't sure of your plans is all - there's no rush so please take your time
